In [1]:
import csv

In [2]:
# 야구단 선수들 현황 읽어들이기
# player_state 라는 리스트(list) 변수를 작성하여 주어진 csv 파일을 읽어서
# 줄 단위로 리스트에 추가하여 전체 현황을 인쇄합니다.
# 단, 인코딩 방식은 UTF-8입니다.

# https://docs.python.org/ko/3.11/library/csv.html

In [3]:
player_state = []


# \ufeff 삭제
# 참고 : https://ddan-dara.tistory.com/44
# 패치 : 반드시 encoding="UTF-8-SIG"로 처리할 것 !

with open('KBO_2023_한국시리즈.csv', newline='', encoding="UTF-8-SIG") as csvfile:
    file_row = csv.reader(csvfile, delimiter=",")
    for row in file_row:
        player_state.append(row)

print(player_state)

print('-------------------------------------------------------------')

[['선수', '1', '2', '3', '4', '5', '6', '7', '8', '9', '타수', '안타', '타점', '득점', '홈런', '볼넷', '삼진', '타율'], ['홍창기', '삼진', '우비', '', '', '우안', '', '중안/2직', '', '4구', '5', '2', '1', '0', '0', '1', '1', '0.294'], ['박해민', '우안', '', '1땅', '', '좌비', '', '좌2', '1땅', '유병', '6', '2', '0', '2', '0', '0', '0', '0.333'], ['김현수', '우홈', '', '유비', '', '우비', '', '우안', '', '', '4', '2', '3', '1', '1', '0', '0', '0.235'], ['최승민', '', '', '', '', '', '', '', '', '', '0', '0', '0', '1', '0', '0', '0', '0'], ['김범석', '', '', '', '', '', '', '', '중안', '', '1', '1', '0', '0', '0', '0', '0', '1'], ['손호영', '', '', '', '', '', '', '', '', '4구', '0', '0', '0', '1', '0', '1', '0', '0'], ['오스틴', '우비', '', '유땅', '', '', '중비', '중안', '', '', '4', '1', '0', '1', '0', '0', '0', '0.375'], ['김민성', '', '', '', '', '', '', '', '4구', '3땅', '1', '0', '0', '1', '0', '1', '0', '0'], ['오지환', '4구', '', '', '2땅', '', '4구', '우홈', '우안', '', '3', '2', '4', '3', '1', '2', '0', '0.4'], ['문보경', '', '우안', '', '3땅', '', '좌홈', '우2', '좌희비', '', '

In [4]:
# 선수, 안타수, 타점, 득점, 홈런, 볼넷, 삼진, 타율 등의 열(column) 인덱스(index)를 구하여 인쇄합니다.
print('선수 : ', player_state[0].index('선수'))
print('안타 : ', player_state[0].index('안타'))
print('타점 : ', player_state[0].index('타점'))
print('득점 : ', player_state[0].index('득점'))
print('홈런 : ', player_state[0].index('홈런'))
print('볼넷 : ', player_state[0].index('볼넷'))
print('삼진 : ', player_state[0].index('삼진'))
print('타율 : ', player_state[0].index('타율'))

선수 :  0
안타 :  11
타점 :  12
득점 :  13
홈런 :  14
볼넷 :  15
삼진 :  16
타율 :  17


---

#### 전체 선수명

In [5]:
player_name = []
#for i in range(1, len(player_state)):
#    player_name.append(player_state[i][player_state[0].index('선수')])

player_name = [player_state[i][player_state[0].index('선수')] for i in range(1, len(player_state))]
print(player_name)

['홍창기', '박해민', '김현수', '최승민', '김범석', '손호영', '오스틴', '김민성', '오지환', '문보경', '정주현', '박동원', '허도환', '문성주', '안익훈', '신민재', '']


---
#### 문1) 선수별 안타수를 추출하여 데이터를 player_hits 리스트에 적재하여 전체 현황을 출력합니다.

In [6]:
player_hits = []
player_hits = [player_state[i][player_state[0].index('안타')] for i in range(1, len(player_state))]
print(player_hits)

['2', '2', '2', '0', '1', '0', '1', '0', '2', '3', '0', '2', '1', '1', '0', '0', '17']


---
#### 문2) 선수별 타점(rbi) 현황을 추출하여 데이터를 player_rbi 리스트에 적재하여 전체 현황을 출력합니다.
#### 선수 타점(RBI:runs batted in) 집계

In [7]:
player_rbi = []
player_rbi = [player_state[i][player_state[0].index('타점')] for i in range(1, len(player_state))]
print(player_rbi)

['1', '0', '3', '0', '0', '0', '0', '0', '4', '3', '0', '0', '1', '2', '0', '1', '15']


---
#### 문3) 선수별 득점 현황을 추출하여 데이터를 player_rs 리스트에 적재합니다.
#### 선수 득점(R:runs scored) 집계

In [8]:
# player_rs = []
# for i in range(1, len(player_state)):
#    player_rs.append(player_state[i][14])

player_rs = [player_state[i][player_state[0].index('득점')] for i in range(1, len(player_state))]
print(player_rs)

['0', '2', '1', '1', '0', '1', '1', '1', '3', '2', '0', '1', '0', '2', '0', '0', '15']


---
#### 문4) 선수별 타율 현황을 추출하여 데이터를 player_batting_avg 리스트에 적재합니다.
#### 선수 타율 집계 => 리스트

In [9]:
# player_batting_avg = []
# for i in range(1, len(player_state)):
#    player_batting_avg.append(float(player_state[i][15]))

player_batting_avg = [player_state[i][player_state[0].index('타율')] for i in range(1, len(player_state))]
print(player_batting_avg)

['0.294', '0.333', '0.235', '0', '1', '0', '0.375', '0', '0.4', '0.462', '0', '0.385', '1', '0.214', '0', '0.188', '']


---
#### 문5) 선수 이름을 키(key)로 하고, 위의 나머지 현황을 tuple로 묶은(zip) 자료구조를 값(value)으로 하는
#### 딕셔너리(dictionary) 자료형의 변수 player_today_state를 생성합니다.

**절차)**


##### 1) 전체적인 구조는 아래와 같은 형식으로 출력되도록 구성해야 됩니다.
##### ex) {'홍창기': ('2', '1', '0', '0.294'), '오지환': ('2', '4', '3', '0.4'), ... (후략)...
#####
##### 2) 먼저 딕셔너리(dict)의 값에 해당되는 부분 (가령, ('2', '1', '3', 0.353) )을 구합니다.
#####    앞선 문제에서 구한 각종 현황(타점, 타율 현황 리스트 등)을 묶어서(zip) 튜플로 변환합니다.
#####    단, 값의 순서는 안타, 타점, 득점, 타율 순으로 배치합니다.
#####
##### 3) 2)번 단계에서 구성한 값을 선수명 리스트를 키(key)로 하여 zip 함수를 이용하여 묶고 이를 dict 함수를
#####   활용하여 딕셔너리 구조로 변환(치환)합니다.

##### 참고) tuple 함수 : https://wikidocs.net/15, https://docs.python.org/ko/3/library/stdtypes.html?highlight=tuple#tuple
##### 참고) zip 함수 : https://wikidocs.net/32#zip, https://docs.python.org/ko/3/library/functions.html#zip
##### 참고) dict 함수 : https://wikidocs.net/16, https://docs.python.org/ko/3/library/stdtypes.html#dict

In [10]:
player_today_state = dict(zip(player_name, tuple(zip(player_hits, player_rbi, player_rs, player_batting_avg))))
print(player_today_state)

{'홍창기': ('2', '1', '0', '0.294'), '박해민': ('2', '0', '2', '0.333'), '김현수': ('2', '3', '1', '0.235'), '최승민': ('0', '0', '1', '0'), '김범석': ('1', '0', '0', '1'), '손호영': ('0', '0', '1', '0'), '오스틴': ('1', '0', '1', '0.375'), '김민성': ('0', '0', '1', '0'), '오지환': ('2', '4', '3', '0.4'), '문보경': ('3', '3', '2', '0.462'), '정주현': ('0', '0', '0', '0'), '박동원': ('2', '0', '1', '0.385'), '허도환': ('1', '1', '0', '1'), '문성주': ('1', '2', '2', '0.214'), '안익훈': ('0', '0', '0', '0'), '신민재': ('0', '1', '0', '0.188'), '': ('17', '15', '15', '')}


---
#### 문6) 타율이 가장 높은 선수의 이름과 타율을 인쇄하십시오.

**절차)**

##### 1) max 함수를 활용하여 최고 타율을 구합니다.
##### 2) 조건문과 반복문 혹은 그 둘의 컴프리헨션을 활용하여 타자의 이름을 구합니다.
##### 3) 반복문을 돌리면서 그 내부에서 조건문을 두어 max 함수로 구한 값과 같은 타자의 이름을 구한 후 분기(break)할 수 있습니다.


In [11]:
# print(max(player_batting_avg))
# best_hitter = ""
#for name in player_today_state.keys():
#    if player_today_state[name][3] == max(player_batting_avg):
#        best_hitter = name
#        break
#print('최고 타율 타자 :', best_hitter, max(player_batting_avg))


print('최고 타율 타자 :', [name for name in player_today_state.keys() if player_today_state[name][3] == max(player_batting_avg)][0]
      , max(player_batting_avg))

최고 타율 타자 : 김범석 1


---
#### 문7) 당일 최다 안타를 친 타자와 안타수

In [12]:
print('당일 최다 안타를 친 타자 :',
      [name for name in player_today_state.keys() if player_today_state[name][0] == max(player_hits)][0]
      , max(player_hits))

당일 최다 안타를 친 타자 : 문보경 3
